### ISSUE TITLE
save_msx_screen error on Yupipati

### ISSUE THREAD

**Issue by sdsnatcher** [openMSX/openMSX#1373]
When I try to use the save_msx_screen command script with the title screen of the game [Yupipati](http://paxangasoft.retroinvaders.com/games/yupipati.html), it gives an "Invalid address" error.

Tested on openMSX-17.0 on macOS-11.5.2

Steps to reproduce:

1) Download the game Yupipati on the link provided above
2) Boot openMSX with the game on a FS-A1GT
3) Wait until the Yupipati title screen with two guys playing paper, rock and scissors to show up, then pause openMSX
4) open the console and type the save_msx_screen command

**Comment by m9710797** [comment]
Hi,

I found that the error originates from this statement:
    puts -nonewline $out [debug read_block $type $addr $size]
on line 141 in <openmsx>/share/scripts/_save_msx_screen.tcl.

The variables $type, $addr, $size have values: VRAM, 0x2f000, 0x800. And [0x2f000,0x2f7ff] is indeed outside the valid range for a turbor machine (with 128kB VRAM).

These values originate from line 114 in the script which looks like:
    lappend sections2 "VRAM" [expr {$spr_pat_base + 0x10000}] 0x800            ;# Sprite character patterns

@sdsnatcher : You originally wrote that code. Can you explain the intention of that line? I think it makes the assumption that $spr_pat_base is in the 1st 64kB and that for the interlaced part we want to look at the 2nd 64kB block? But is that assumption correct for sprite patterns and sprite attributes? Or maybe (in case of interlace) for the first part we need to clear the top address bit and for the second part we need to set it ???

In other words: I know why the error triggers. But I don't know how the script is supposed to work in this case. So I need your help.

**Comment by NYYR1KK1** [comment]
> I think it makes the assumption that $spr_pat_base is in the 1st 64kB and that for the interlaced part we want to look at the 2nd 64kB block? But is that assumption correct for sprite patterns and sprite attributes?

I think this is the best assumption you can make... but clearly there is a design error in the code. All these "+ 0x10000", "+ 0x8000", "- (0x8000 * $interlace)" and "- (0x10000 * $interlace)" should be changed to "^ 0x10000", "^ 0x8000", "^ (0x8000 * $interlace)" and "^ (0x10000 * $interlace)"

**Comment by m9710797** [comment]
Thanks Nyyr1kk1. I'm sure that'll fix the problem. I'll try it this evening.

But still I'd like to better understand the intention of the script:
* I perfectly understand that (for interlaced images) you want to save both pages for the *bitmap* data. But I don't immediately see why you want to do the same for the *sprite* patterns and attributes.  Also why save the *palette* data from both pages?
* Saving too much doesn't hurt of course. But I'd like to understand if this was an oversight (just copy/pasted the original code for the 2nd page). Or if it really was intended like this.

**Comment by sdsnatcher** [comment]
Quick reply since I'm short on time.

This part of the script was highly experimental:

1) My degree of intimacy with TCL is minimal, so I'm sure some parts could have been better written by someone more experienced. I just took the approach "something is better than nothing", so optimizations are welcome.

2) I remember someone doing tests with interlaced sprites back then on the MRC forums. IIRC, you have to flip the sprite tables by software but it was doable. This additional data would have to be manually copied though by the guy who dumped the image, but the space for it would be reserved.  But if you feel that this is unnecessary, feel free optimize that part too. I just recommend to test the results with the native MSX viewers like BLS.COM.

**Comment by NYYR1KK1** [comment]
What ever you save with the other bitmap is just a lucky guess that might or might not contain some valid information... I'm starting to think that as it is such a shot in the dark, it might be more reasonable just to save the bitmap as that is the only thing that we really know to exist.

**Comment by sdsnatcher** [comment]
Ok. One idea might be to comment the parts of the code that deal with the palette and sprites on the 2nd page, so if someone needs that for some specific case, it will be easy to just uncomment.

**Comment by m9710797** [comment]
I've implemented Nyyr1kk1's suggestion: use xor rather than add/sub for address calculations involving 'interlace'.
I've not changed anything else. I'll leave that to one of you to experiment with.

For now I'm closing this ticket. But feel free to reopen (or create a new one) if you still want to experiment with simplifying the script.
